File: 3475-minimum-operations-to-make-binary-array-elements-equal-to-one-i/3475-minimum-operations-to-make-binary-array-elements-equal-to-one-i.cpp
class Solution {
public:
    int minOperations(vector<int>& nums) {
        int n = nums.size();
        int ans = 0;
        vector<int> nums1 = nums;
        for(int right=0; right<n-2; right++){
            if(nums1[right]==0){
                nums1[right] = 1;
                nums1[right+1] ^= 1;
                nums1[right+2] ^= 1;
                ans++;
            }
        }
        for(int right = n-3; right<n; right++){
            if(nums1[right]==0){
                return -1;
            }
        }
        return ans;
    }
};

File: 3475-minimum-operations-to-make-binary-array-elements-equal-to-one-i/3475-minimum-operations-to-make-binary-array-elements-equal-to-one-i.rs
impl Solution {
    pub fn min_operations(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut ans = 0;
        let mut nums1 = nums.clone();

        for right in 0.. (n-2) {
            if nums1[right] == 0 {
//Declaring something as mut allows modification without taking ownership.
//Acessing an element using usize still gives us a &i32(reference) only.
//But, Rust's "Deref Coercion" automatically applies '*' (de-reference) where necessary for assignments and operations.
//Since we explicitly declared the value as mut, Rust knows we intend to modify it.
//Also, assignment & arithmetic & also comparision triggers auto-dereferencing.
                nums1[right] = 1;
                nums1[right + 1] ^= 1;
                nums1[right + 2] ^= 1;
                ans+=1;
            }
        }

        if (nums1[n-1]==0) || (nums1[n-2]==0) {
            return -1;
        }
        ans
    }
}